### PR TITLE
[sdk/dotnet] Fix async await warnings

### LINKED ISSUE #7422: [sdk/dotnet] Possible race condition results in `NullReferenceException`
Recently I've been experiencing random failures with `NullReferenceException` which looks due to a race condition:
```
Diagnostics:
  pulumi:pulumi:Stack (aws-eks-alpha):
    error: Running program 'D:\Devel\Mps\devops-gemini-pulumi\Gemini\bin\Debug\gemini.dll' failed with an unhandled exception:
    System.NullReferenceException: Object reference not set to an instance of an object.
       at Task<HashSet<string>> Pulumi.Deployment.GetAllTransitivelyReferencedCustomResourceUrnsAsync(HashSet<Resource> resources)+(CustomResource r) => { } in /_/sdk/dotnet/Pulumi/Deployment/Deployment_Prepare.cs:line 152
       at bool System.Linq.Enumerable+SelectEnumerableIterator<TSource, TResult>.MoveNext()
       at Task<TResult[]> System.Threading.Tasks.Task.WhenAll<TResult>(IEnumerable<Task<TResult>> tasks)
       at async Task<HashSet<string>> Pulumi.Deployment.GetAllTransitivelyReferencedCustomResourceUrnsAsync(HashSet<Resource> resources) in /_/sdk/dotnet/Pulumi/Deployment/Deployment_Prepare.cs:line 153
       at async Task<PrepareResult> Pulumi.Deployment.PrepareResourceAsync(string label, Resource res, bool custom, bool remote, ResourceArgs args, ResourceOptions options) in /_/sdk/dotnet/Pulumi/Deployment/Deployment_Prepare.cs:line 80
       at async Task<(string urn, string id, Struct data, ImmutableDictionary<string, ImmutableHashSet<Resource>> dependencies)> Pulumi.Deployment.RegisterResourceAsync(Resource resource, bool remote, Func<string, Resource> newDependency, ResourceArgs args, ResourceOptions options) in /_/sdk/dotnet/Pulumi/Deployment/Deployment_RegisterResource.cs:line 27
       at async Task<(string urn, string id, Struct data, ImmutableDictionary<string, ImmutableHashSet<Resource>> dependencies)> Pulumi.Deployment.ReadOrRegisterResourceAsync(Resource resource, bool remote, Func<string, Resource> newDependency, ResourceArgs args, ResourceOptions options) in /_/sdk/dotnet/Pulumi/Deployment/Deployment_ReadOrRegisterResource.cs:line 73
       at async Task Pulumi.Deployment.CompleteResourceAsync(Resource resource, bool remote, Func<string, Resource> newDependency, ResourceArgs args, ResourceOptions options, ImmutableDictionary<string, IOutputCompletionSource> completionSources) in /_/sdk/dotnet/Pulumi/Deployment/Deployment_ReadOrRegisterResource.cs:line 89
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass10_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) x 2 in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 137
       at async Task<int> Pulumi.Deployment+Runner.WhileRunningAsync() in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 177
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass10_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) x 2 in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 137
       at void System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```
```
Diagnostics:
  pulumi:pulumi:Stack (k8s-alpha):
    error: Running program 'D:\Devel\Mps\devops-gemini-pulumi\Gemini\bin\Debug\gemini.dll' failed with an unhandled exception:
    System.NullReferenceException: Object reference not set to an instance of an object.
       at Output<T> Pulumi.Input<T>.op_Implicit(Input<T> input) in /_/sdk/dotnet/Pulumi/Core/Input.cs:line 43
       at async Task<OutputData<ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>>> Pulumi.Output<T>.TupleHelperAsync<T1, T2, T3, T4, T5, T6, T7, T8>(Input<T1> item1, Input<T2> item2, Input<T3> item3, Input<T4> item4, Input<T5> item5, Input<T6> item6, Input<T7> item7, Input<T8> item8)+GetData(?) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 328
       at async Task<OutputData<ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>>> Pulumi.Output<T>.TupleHelperAsync<T1, T2, T3, T4, T5, T6, T7, T8>(Input<T1> item1, Input<T2> item2, Input<T3> item3, Input<T4> item4, Input<T5> item5, Input<T6> item6, Input<T7> item7, Input<T8> item8) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 318
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<OutputData<U>> Pulumi.Output<T>.ApplyHelperAsync<U>(Task<OutputData<T>> dataTask, Func<T, Output<U>> func) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 254
       at async Task<bool> Pulumi.Utilities.OutputUtilities.GetIsKnownAsync<T>(Output<T> output) in /_/sdk/dotnet/Pulumi/Core/OutputUtilities.cs:line 53
       at async Task<ImmutableHashSet<Resource>> Pulumi.Output<T>.Pulumi.IOutput.GetResourcesAsync() in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 159
       at async Task<OutputData<ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>>> Pulumi.Output<T>.TupleHelperAsync<T1, T2, T3, T4, T5, T6, T7, T8>(Input<T1> item1, Input<T2> item2, Input<T3> item3, Input<T4> item4, Input<T5> item5, Input<T6> item6, Input<T7> item7, Input<T8> item8)+GetData(?) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 329
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) x 2 in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<OutputData<ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>>> Pulumi.Output<T>.TupleHelperAsync<T1, T2, T3, T4, T5, T6, T7, T8>(Input<T1> item1, Input<T2> item2, Input<T3> item3, Input<T4> item4, Input<T5> item5, Input<T6> item6, Input<T7> item7, Input<T8> item8) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 314
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<OutputData<U>> Pulumi.Output<T>.ApplyHelperAsync<U>(Task<OutputData<T>> dataTask, Func<T, Output<U>> func) x 2 in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 254
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<OutputData<object>> Pulumi.Output<T>.Pulumi.IOutput.GetDataAsync() in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 164
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<object> Pulumi.Serialization.Serializer.SerializeAsync(string ctx, object prop, bool keepResources) in /_/sdk/dotnet/Pulumi/Serialization/Serializer.cs:line 138
       at async Task<SerializationResult> Pulumi.Deployment.SerializeFilteredPropertiesAsync(string label, IDictionary<string, object> args, Predicate<string> acceptKey, bool keepResources) in /_/sdk/dotnet/Pulumi/Deployment/Deployment_Serialization.cs:line 53
```
```
Diagnostics:
  pulumi:pulumi:Stack (k8s-alpha):
    error: Running program 'D:\Devel\Mps\devops-gemini-pulumi\Gemini\bin\Debug\gemini.dll' failed with an unhandled exception:
    System.NullReferenceException: Object reference not set to an instance of an object.
       at Output<T> Pulumi.Input<T>.op_Implicit(Input<T> input) in /_/sdk/dotnet/Pulumi/Core/Input.cs:line 43
       at async Task<OutputData<ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>>> Pulumi.Output<T>.TupleHelperAsync<T1, T2, T3, T4, T5, T6, T7, T8>(Input<T1> item1, Input<T2> item2, Input<T3> item3, Input<T4> item4, Input<T5> item5, Input<T6> item6, Input<T7> item7, Input<T8> item8)+GetData(?) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 328
       at async Task<OutputData<ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>>> Pulumi.Output<T>.TupleHelperAsync<T1, T2, T3, T4, T5, T6, T7, T8>(Input<T1> item1, Input<T2> item2, Input<T3> item3, Input<T4> item4, Input<T5> item5, Input<T6> item6, Input<T7> item7, Input<T8> item8) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 318
       at async Task<OutputData<U>> Pulumi.Output<T>.ApplyHelperAsync<U>(Task<OutputData<T>> dataTask, Func<T, Output<U>> func) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 254
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) x 2 in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<bool> Pulumi.Utilities.OutputUtilities.GetIsKnownAsync<T>(Output<T> output) in /_/sdk/dotnet/Pulumi/Core/OutputUtilities.cs:line 53
       at async Task<OutputData<ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>>> Pulumi.Output<T>.TupleHelperAsync<T1, T2, T3, T4, T5, T6, T7, T8>(Input<T1> item1, Input<T2> item2, Input<T3> item3, Input<T4> item4, Input<T5> item5, Input<T6> item6, Input<T7> item7, Input<T8> item8) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 314
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) x 2 in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<OutputData<U>> Pulumi.Output<T>.ApplyHelperAsync<U>(Task<OutputData<T>> dataTask, Func<T, Output<U>> func) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 254
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<OutputData<U>> Pulumi.Output<T>.ApplyHelperAsync<U>(Task<OutputData<T>> dataTask, Func<T, Output<U>> func) in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 254
       at async void Pulumi.Deployment+Runner+<>c__DisplayClass13_0.<WhileRunningAsync>g__HandleCompletion|0(?)+HandleCompletion(?) in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 141
       at async Task<OutputData<object>> Pulumi.Output<T>.Pulumi.IOutput.GetDataAsync() in /_/sdk/dotnet/Pulumi/Core/Output.cs:line 164
       at void System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```
I haven't had time to dig into them yet. As usual I'm using the very latest version of Pulumi & Pulumi.Automation (3.6.0), although the above were captured just before upgrading.

### PR DESCRIPTION
**Pull request by gitfool**
# Description

Fixes async await warnings found while looking into #7422. (Does not fix #7422 itself.)

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by gitfool** [comment]
@t0yv0 PTAL

File: sdk/dotnet/Pulumi.Automation.Tests/LocalWorkspaceTests.cs [hunk 1]
```diff
@@ -1659,12 +1659,12 @@ public async Task InlineProgramExceptionDuringUpShouldNotDeleteResources()
...
-            var config = new Dictionary<string, ConfigValue>()
-            {
-                ["ShouldFail"] = new ConfigValue("false"),
-            };
             try
             {
+                var config = new Dictionary<string, ConfigValue>
+                {
+                    ["ShouldFail"] = new ConfigValue("false"),
+                };
```
Comment (mikhailshilkov) [inline sdk/dotnet/Pulumi.Automation.Tests/LocalWorkspaceTests.cs:1667]: I'm curious why this is better

File: sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs [hunk 2]
```diff
@@ -518,14 +518,14 @@ public override async Task<ConfigValue> GetConfigAsync(string stackName, string
         }
 
         /// <inheritdoc/>
-        public override async Task SetConfigAsync(string stackName, string key, ConfigValue value, CancellationToken cancellationToken = default)
+        public override Task SetConfigAsync(string stackName, string key, ConfigValue value, CancellationToken cancellationToken = default)
         {
             var secretArg = value.IsSecret ? "--secret" : "--plaintext";
-            await this.RunCommandAsync(new[] { "config", "set", key, value.Value, secretArg, "--stack", stackName }, cancellationToken).ConfigureAwait(false);
+            return this.RunCommandAsync(new[] { "config", "set", key, value.Value, secretArg, "--stack", stackName }, cancellationToken);
```
Comment (mikhailshilkov) [inline sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs:524]: Is this done by resharper? Two versions aren't exactly the same, there are some subtle difference in exception propagation, but I think it shouldn't matter for us here. Just interesting if the tool is confident to suggest this.

File: sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs [hunk 3]
```diff
@@ -19,13 +19,13 @@ public GrpcEngine(string engine)
             this._engine = new Engine.EngineClient(new Channel(engine, ChannelCredentials.Insecure, grpcChannelOptions));
         }
         
-        public async Task LogAsync(LogRequest request)
-            => await this._engine.LogAsync(request);
+        public Task LogAsync(LogRequest request)
+            => this._engine.LogAsync(request).ResponseAsync;
```
Comment (mikhailshilkov) [inline sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs:23]: What is `ResponseAsync`?

File: sdk/dotnet/Pulumi.Automation.Tests/LocalWorkspaceTests.cs (hunk 1 above)
Comment (gitfool) [reply sdk/dotnet/Pulumi.Automation.Tests/LocalWorkspaceTests.cs:1667]: Moving the `config` var changes the block scope and fixes the Resharper issue "Local variable 'config' hides outer local variable with the same name" for line 1641.

File: sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs (hunk 2 above)
Comment (gitfool) [reply sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs:524]: No, that's not coming from Resharper. I installed the latest versions of the analyzers suggested [here](https://cezarypiatek.github.io/post/async-analyzers-summary/):
```xml
<ItemGroup>
  <PackageReference Include="AsyncFixer" Version="1.5.1">
    <PrivateAssets>all</PrivateAssets>
    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
  </PackageReference>
  <PackageReference Include="Asyncify" Version="0.9.7" >
    <PrivateAssets>all</PrivateAssets>
    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
  </PackageReference>
  <PackageReference Include="Meziantou.Analyzer" Version="1.0.667">
    <PrivateAssets>all</PrivateAssets>
    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
  </PackageReference>
  <PackageReference Include="Microsoft.VisualStudio.Threading.Analyzers" Version="16.10.56" >
    <PrivateAssets>all</PrivateAssets>
    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
  </PackageReference>
</ItemGroup>
```
This particular warning is "AsyncFixer01: The method `SetConfigAsync` does not need to use async/await."

File: sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs (hunk 3 above)
Comment (gitfool) [reply sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs:23]: This has the same warning AsyncFixer01 since the await is effectively redundant.

The return type is `AsyncUnaryCall<TResponse>` which is awaitable via its `GetAwaiter` method:
```c#
public TaskAwaiter<TResponse> GetAwaiter()
{
    return responseAsync.GetAwaiter();
}
```
The `responseAsync` field holds the `Task` that can be awaited:
```c#
readonly Task<TResponse> responseAsync;
```
And this field is otherwise accessible via the `ResponseAsync` property:
```c#
public Task<TResponse> ResponseAsync
{
    get
    {
        return this.responseAsync;
    }
}
```
So to remove the redundant await I return the underlying task via this property.

File: sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs (hunk 2 above)
Comment (gitfool) [reply sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs:524]: Normally there's no need to await unless it's a synchronization point after which something else needs to be executed. In this case it's a tail call in the function and clearly nothing follows it at that scope, so awaiting can be deferred to an outer scope.

Comment (gitfool) [reply sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs:524]: Capturing the caveats:
* [Redundant async/await](https://cezarypiatek.github.io/post/async-analyzers-p1/#1-redundant-asyncawait)
* [Eliding Async and Await](https://blog.stephencleary.com/2016/12/eliding-async-await.html)

TL;DR: if in doubt we can leave them in to avoid possible subtle changes in behavior.

**Comment by mikhailshilkov** [comment]
I'm getting test failures here, could you please try them locally? https://github.com/pulumi/pulumi/runs/3122575299?check_suite_focus=true

**Comment by gitfool** [comment]
Will do.

**Comment by gitfool** [comment]
@mikhailshilkov try again.

**Review by mikhailshilkov — APPROVED** [review]
LGTM. Could you please fix the changelog conflict and I'll merge?

**Comment by gitfool** [comment]
@mikhailshilkov squashed, rebased and force pushed.

**Comment by t0yv0** [comment]
Looking good! Sorry this somehow slipped out of my queue. 

Can you let me know which tool you used to analyze this? I can maybe rig it up into our CI.

File: sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs (hunk 2 above)
Comment (gitfool) [reply sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs]: @t0yv0 re tooling see the analyzers above. I'm not sure I'd recommend integrating them more permanently though as they spew other suggestions / warnings which I was ignoring to focus on async await.

**Review by t0yv0 — COMMENTED** [review]
LGTM
